File: hamiltonian_cycle.c
#include <stdio.h>
#include <stdlib.h>
#include <conio.h>
#define MAX_V 10  

int V;                                            
int graph[MAX_V][MAX_V];         

void printSolution(int path[]) {
    int i ;
    printf("Hamiltonian Cycle: \n");
    for ( i = 0; i < V; i++)
        printf("%d ", path[i]);
    printf("%d\n", path[0]);                        
}

int isSafe(int v, int path[], int pos) {
    int i;
    if (graph[path[pos - 1]][v] == 0)
        return 0;                         

    for ( i = 0; i < pos; i++)
        if (path[i] == v)
            return 0;                       

    return 1;
}

int hamCycleUtil(int path[], int pos) {
    int count,v;
    if (pos == V) {
        if (graph[path[pos - 1]][path[0]] == 1) {
            printSolution(path);                               
            return 1;                                                   
        }
        return 0;
    }

     count = 0;                                      
    for ( v = 1; v < V; v++) {
        if (isSafe(v, path, pos)) {
            path[pos] = v;
            count += hamCycleUtil(path, pos + 1);                   

            path[pos] = -1;            
        }
    }
    return count;
}

void hamCycle() {
    int i;
    int cycleCount;
    int path[MAX_V];
    for (i = 0; i < V; i++)
        path[i] = -1;

    path[0] = 0; 
    cycleCount = hamCycleUtil(path, 1);

    if (cycleCount == 0)
        printf("No Hamiltonian Cycle exists\n");
    else
        printf("Total Hamiltonian Cycles found: %d\n", cycleCount);
}

int main() {
    int i , j;
    // clrscr();
    printf("Enter the number of vertices: ");
    scanf("%d", &V);

    printf("Enter the adjacency matrix:\n");
    for ( i = 0; i < V; i++) {
        for (j = 0; j < V; j++) {
            scanf("%d", &graph[i][j]);
        }
    }

    hamCycle();
    // getch();
    return 0;
}

File: hamiltonian_cycle.java
import java.util.Scanner;

public class hamiltonian_cycle {
    static int max =10;
    static int V;
    static int g[][] = new int[max][max];
    static void print(int path[])
    {
        System.out.println("Cycle: ");
        for(int i=0; i<V; i++)
        {
            System.out.print(""+path[i]+" ");
        }
        System.out.print(""+path[0]);
        System.out.println();
    }
    static boolean safe(int path[], int pos, int v)
    {
        if(g[path[pos-1]][v] == 0)
            return false;
        for(int i=0; i<V; i++)
        {
            if(path[i] == v)
                return false;
        }
        return true;
    }
    static int cycle(int pos, int path[])
    {
        int count, v;
        if(pos == V)
        {
            if(g[path[pos-1]][path[0]] == 1)
                {
                    print(path);
                    return 1;
                }
            else 
                return 0;
        }
        count = 0;
        for(v=0; v<V; v++)
        {
            if(safe(path, pos, v))
            {
                path[pos] = v;
                count += cycle(pos+1, path);
                path[pos]=-1;
            }
        }
        return count;
    }
    static void hami(){
        int count;
        int path[] = new int[V];
        for(int i=0; i<V; i++)
        {
            path[i]=-1;
        }
        path[0] = 0;

        count = cycle(1, path);
        if(count == 0)
        {
            System.out.println("fail");
        }
        else
        {
            System.out.println(count);
        }
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Vertices:");
        V = sc.nextInt();
        for(int i=0; i<V; i++)
            for(int j=0; j<V; j++)
                g[i][j]=sc.nextInt();

        hami();
        sc.close();
    }
}
